packages/playwright/src/fixture.ts

```diff
@@ -132,21 +133,140 @@ type Playwright = {
   _instrumentation: ClientInstrumentation;
 };
 
+let wsDeferred: Deferred<WebSocket> | undefined;
+
+async function waitUntilConnected() {
+  if (wsDeferred) {
+    return wsDeferred.promise;
+  }
+  const deferred = (wsDeferred = createDeferred());
+
+  const port = getServerPort();
+  const ws = new WebSocket(`ws://localhost:${port}`);
+
+  ws.on("open", () => deferred.resolve(ws));
+  ws.on("error", error => deferred.reject(error));
+  // TODO: close WS connections on test end
+  // to avoid relying on TestInfo's internals for this the reporter could notify the connection when the test ends
+
+  return wsDeferred.promise;
+}
+
+const fixtureStates = new WeakMap<
+  TestInfoInternal,
+  {
+    expectSteps: Set<string>;
+    ignoredSteps: Set<string>;
+    testIdData: TestExecutionIdData;
+  }
+>();
+
+function getFixtureState(testInfo: TestInfoInternal) {
+  let state = fixtureStates.get(testInfo);
+  if (!state) {
+    const testIdData: TestExecutionIdData = {
+      filePath: testInfo.file,
+      projectName: testInfo.project.name,
+      repeatEachIndex: testInfo.repeatEachIndex,
+      attempt: testInfo.retry + 1,
+      source: {
+        title: testInfo.title,
+        // this one only drops the filename (first segment) and the test title (last segment)
+        // it's different from the one in the reporter, since the "root" suites are just the file suites created here:
+        // https://github.com/microsoft/playwright/blob/a6488c4a2879a22e8da0f6708114ef7b9f4d253f/packages/playwright/src/common/testLoader.ts#L36
+        // in this context they are not attached to the root and project suites
+        scope: testInfo.titlePath.slice(1, -1),
+      },
+    };
+    state = {
+      expectSteps: new Set(),
+      ignoredSteps: new Set(),
+      testIdData,
+    };
+    fixtureStates.set(testInfo, state);
+  }
+  return state;
+}
+
 const patchedTestInfos = new WeakSet<TestInfoInternal>();
 
 export async function replayFixture(
   { playwright, browser }: { playwright: Playwright; browser: Browser },
   use: () => Promise<void>,
   testInfo: TestInfoInternal
 ) {
+  const { expectSteps, ignoredSteps, testIdData } = getFixtureState(testInfo);
+
   // fixtures are created repeatedly for the same test
-  // since they are often created for hooks too
-  // it's important to avoid setting up the fixture multiple times
+  // since they are created for beforeAll and afterAll hooks too
+  // it's important to avoid patching the test info multiple times
   // otherwise the reporter would get notified multiple times about the same steps
-  if (patchedTestInfos.has(testInfo)) {
-    return use();
```

Comment:
Essentially, this PR ensures that we attach the instrumentation listener in this scenario while avoiding patching the test info twice.

To do that I introduced cached `getFixtureState` since we need to use the same `expectSteps` and `ignoredSteps` sets across doubled fixture runs like this.

And I made sure that we only create a single WS connection here to spare some resources and that both of those fixture runs await the connection properly. That said the WS code will hopefully go away from this fixture soon - I'm working on that separately.